### PR TITLE
Allow some lazy loops to upgrade with auto-atomicity

### PR DESCRIPTION
**Pull request by stephentoub**
For example, today we will convert `a*b` into the equivalent of `(?>a*)b`, but we won't modify `a*?b`... with this PR, that will also become the equivalent of `(?>a*)b`.

### CODE REVIEW

**Comment by joperezr** [comment]
>with this PR, that will also become the equivalent of `(?>a*)b`.

I suppose you meant the equivalent of `(?>a*?)b`?

**Comment by stephentoub** [comment]
> I suppose you meant the equivalent of `(?>a*?)b`?

No :-)

**Comment by joperezr** [comment]
Yes sorry, that makes sense, I had a brain freeze for a second and missed the point of the  PR 😆

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs [hunk 1]
```diff
@@ -1807,7 +1807,29 @@ static void ProcessNode(RegexNode node, RegexNode subsequent)
...
+
+                        case RegexNodeKind.Onelazy or RegexNodeKind.Notonelazy or RegexNodeKind.Setlazy when CanBeMadeAtomic(node, subsequent, iterateNullableSubsequent: false, allowLazy: true):
+                            // The lazy loop doesn't overlap with what comes after it, which means it needs to match as much as its allowed
+                            // to match in order for there to be a possibility that what comes next matches (if it doesn't match as much
+                            // as it's allowed and there was still more it could match, then what comes next is guaranteed to not match,
+                            // since it doesn't match any of the same things the loop matches).  We don't want to just make the lazy loop
+                            // atomic, as an atomic lazy loop matches as little as possible, not as much as possible.  Instead, we want to
+                            // make the lazy loop into an atomic greedy loop.  Note that when we check CanBeMadeAtomic, we need to set
+                            // "iterateNullableSubsequent" to false so that we only inspect non-nullable subsequent nodes.  For example,
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:1826]: Thanks for adding the comments, they are very helpful. Should we also add a line in there for why one uses allowLazy=true and the other one false?

**Review by joperezr — APPROVED** [review]

**Pull request merged**